runtime/lua/vim/diagnostic.lua

```diff
@@ -658,7 +658,7 @@ function M.set(namespace, bufnr, diagnostics, opts)
 
   vim.api.nvim_buf_call(bufnr, function()
     vim.api.nvim_command(
-      string.format("doautocmd <nomodeline> DiagnosticChanged %s", vim.api.nvim_buf_get_name(bufnr))
+      string.format("doautocmd <nomodeline> DiagnosticChanged %s", vim.api.nvim_buf_get_name(bufnr):gsub("|", "\\|"))
```

Comment:
Maybe using `vim.fn.fnameescape()` is better.
```suggestion
      string.format("doautocmd <nomodeline> DiagnosticChanged %s", vim.fn.fnameescape(vim.api.nvim_buf_get_name(bufnr)))
```

Reply:
Thank you, fixed in c52aa7c.